### PR TITLE
Pass additional model params to the LLM

### PR DESCRIPTION
**Pull request by malexw**
## Problem

Currently `canopy-server` and the `LLM` model params are all pre-configured. In order to offer more flexibility in how the underlying LLM is used, we'd like to enhance `canopy-server` to forward the params that are passed with each `/chat/completion` call to the underlying `LLM`.

## Solution

This PR addresses issue #164. We add two new config parameters, `allow_api_key_override` and `allow_model_params_override`, with both defaulting to `true`.

When `allow_api_key_override` is true, Canopy will read the API key from the request's `Authorization` header, if set, and forward it on to the LLM to be used in API calls. If no `Authorization` header is set or the config value is false, Canopy continues to use the api key specified by the config or env var.

When `allow_model_params_override` is true, any model parameters specified in the API call are forwarded to the LLM. For example, in the OpenAI API, these might be `temperature`, `top_p`, or `logit_bias`. Two parameters that aren't currently forwarded are `tools` and `tool_choice`, as well as the deprecated `functions` and `function_call`. Getting Canopy to work with function calling would require a bit of additional work that I haven't done here.

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update
- [ ] Infrastructure change (CI configs, etc)
- [ ] Non-code change (docs, etc)
- [ ] None of the above: (explain here)

## Test Plan

Tests were run manually to check that parameters are getting passed through and that Canopy properly handles the response coming back (for example, that all the choices are returned for `n` > 1). Two unit tests were also added one for the `override` config values being `False`, and one for them being `True`.

### CODE REVIEW

File: src/canopy/llm/openai.py [hunk 1]
```diff
@@ -94,9 +94,11 @@ def chat_completion(self,
         model_params_dict.update(
             model_params or {}
         )
+        model = model_params.get("model", self.model_name) if model_params else self.model_name
+        model_params_dict = {k: v for k, v in model_params_dict.items() if k != 'model' and v is not None} if model_params else None
```
Comment (malexw) [inline src/canopy/llm/openai.py:98]: I think I more complete solution here will probably need to make sure we're only passing params to OpenAI that their API supports

File: src/canopy/chat_engine/chat_engine.py [hunk 2]
```diff
@@ -197,8 +197,10 @@ def chat(self,
             messages,
             max_tokens=self.max_prompt_tokens
         )
+        max_tokens = model_params.get("max_tokens", self.max_generated_tokens) if model_params else self.max_generated_tokens
+        model_params = {k: v for k, v in model_params.items() if k != 'max_tokens'} if model_params else None
         llm_response = self.llm.chat_completion(llm_messages,
```
Comment (malexw) [inline src/canopy/chat_engine/chat_engine.py:202]: Should we be updating `BaseLLM` to move the `max_tokens` and `stream` param inside of `model_params`?

Or should we go the other way and pull more params out of `model_params`? Or neither?

File: src/canopy/llm/openai.py [hunk 3]
```diff
@@ -91,15 +93,19 @@ def chat_completion(self,
...
-        model_params_dict.update(
-            model_params or {}
-        )
+        if self.request_params_passthrough:
+            passthrough_params = {
+                k: v for k, v in (model_params or {}).items() if v is not None
+            }
+            model_params_dict.update(passthrough_params)
+        if model_params_dict.get("model", None) is None:
+            model_params_dict["model"] = self.model_name
```
Comment (igiloh-pinecone) [inline src/canopy/llm/openai.py:102]: You can probably do that already on `__init()__`:
```python
        super().__init__(model_name)
        ...
        self.default_model_params = kwargs
        self.default_model_params["model] = self.model_name
```
And then the `update()` above would suffice

**Review by igiloh-pinecone — CHANGES_REQUESTED** [review]
@malexw Please see one suggested change - move the passthrough control from the `LLM` level to the `ChatEngine` level.

File: src/canopy/llm/openai.py (hunk 3 above)
Comment (malexw) [reply src/canopy/llm/openai.py]: Good suggestion!

File: config/config.yaml [hunk 4]
```diff
@@ -50,7 +52,7 @@ chat_engine:
     params:
       model_name: gpt-3.5-turbo         # The name of the model to use.
       # You can add any additional parameters which are supported by the LLM's `ChatCompletion()` API. The values set
-      # here will be used in every LLM API call. For example:
+      # here will be used in every LLM API call unless `request_params_passthrough` is false. For example:
```
Comment (igiloh-pinecone) [inline config/config.yaml:55]:
```suggestion
      # here will be used in every LLM API call unless `allow_model_params_override` is false. For example:
```

**Comment by malexw** [comment]
Sure thing @igiloh-pinecone - I've updated this one to be just for param passing.

The API key changes part of this PR didn't feel right to me either, but without doing deeper rearchitecting I'm not sure that we had a lot of options. Should I hold back on the API key overriding PR while we think of a better way to do it? Also happy to jump on a call if it helps to talk things through. My contact info is in my profile.

**Review by igiloh-pinecone — COMMENTED** [review]
All in all LGTM!  
See two last comments

**Comment by malexw** [comment]
Great, I hope it's good to go now, @igiloh-pinecone 

In the config file, I ended up writing something a little bit different from what you recommended. When I looked at it with fresh eyes today, I felt my original wording could be a little clearer. Definitely let me know if you want further changes.

**Review by igiloh-pinecone — APPROVED** [review]
@malexw thank you for all of your efforts and apologies for the late response!

LGTM! PR is approved.  
Note that there are quite a few conflict, due to a recent refactor we've introduced in #231. If you'd like, someone from our end can take care of the conflicts - you've already invested a lot of effort into this PR.

**Comment by malexw** [comment]
@igiloh-pinecone Sounds good! I've attempted to resolve the merge myself, but I think there might be an issue around how token budgets are calculated in `chat_engine.py`. I think the calculations work when we assume that `max_tokens` is set once in the config file, but needs to be done differently if we allow for it to be overridden. I think your team probably has a better vision for how to resolve that, so I'll hand it over to you.

**Comment by igiloh-pinecone** [comment]
I think the calculations work when we assume that `max_tokens` is set once in the config file, but needs to be done differently if we allow for it to be overridden.

Thanks @malexw!!  
I've taken a deeper look - the calculations you're talking about are only affected by Canopy's built-in `max_prompt_tokens` (== input tokens sent to the model), not the `max_generated_tokens` (== tokens that the model is allowed to generate).  
At runtime, users could only affect the _generated_ tokens, by sending the `max_token` parameter - which doesn't affect Canopy's behavior.

Other than that, I'm about to approve and merge this PR.  
Thank you again very much for your efforts and contribution!!

File: src/canopy/llm/openai.py [hunk 5]
```diff
@@ -117,10 +119,8 @@ def chat_completion(self,
         messages = [SystemMessage(content=system_message).dict()
                     ] + [m.dict() for m in chat_history]
         try:
-            response = self._client.chat.completions.create(model=self.model_name,
```
Comment (igiloh-pinecone) [inline src/canopy/llm/openai.py:120]: @malexw I think it's better to pass the `model=` explicitly, for code readability.  
I'm doing a small change around this and pushing to your branch.

**Review by igiloh-pinecone — APPROVED** [review]
LGTM!

File: tests/unit/chat_engine/test_chat_engine.py [hunk 6]
```diff
@@ -130,18 +130,28 @@ def test_chat(self, namespace, history_length=5, snippet_length=10):
...
-            model_params=None
+            model_params={'max_tokens': 200}
         )
 
         assert response == expected['response']
 
     @pytest.mark.parametrize("namespace", [
         None, TEST_NAMESPACE
     ])
+    @pytest.mark.parametrize("allow_model_params_override,params_override",
```
Comment (igiloh-pinecone) [inline tests/unit/chat_engine/test_chat_engine.py:142]: @malexw I've removed the copy-pasted test cases, making the existing test parameterized instead.

**Pull request merged**